core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -63,7 +67,18 @@ public abstract class AbstractScoreDirector<Solution_, Score_ extends Score<Scor
     private long calculationCount = 0L;
     protected Solution_ workingSolution;
     protected Integer workingInitScore = null;
-    private ShadowVariablesAssert beforeMoveSnapshot;
+    private AllVariablesAssert<Solution_> beforeMoveSnapshot;
+    private Solution_ beforeMoveSolution;
+    private AllVariablesAssert<Solution_> afterMoveSnapshot;
+    private Solution_ afterMoveSolution;
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> beforeVariableChangedForwardEvents = new ArrayList<>();
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> afterVariableChangedForwardEvents = new ArrayList<>();
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> beforeVariableChangedUndoEvents = new ArrayList<>();
+    private final List<Pair<VariableDescriptor<Solution_>, Object>> afterVariableChangedUndoEvents = new ArrayList<>();
+
+    // These points to either ...ForwardEvents or ...UndoEvents, or null if not in an assert mode
+    private List<Pair<VariableDescriptor<Solution_>, Object>> beforeVariableChangedEvents = null;
+    private List<Pair<VariableDescriptor<Solution_>, Object>> afterVariableChangedEvents = null;
```

Comment:
This complicates the score director far too much.

Let's think of a way of encapsulating it in a single type. All of this work should happen within that type, with the interactions between `ScoreDirector` and the other type being kept at a minimum;  thinking out loud here, perhaps some event-based system could do.

Once we have this in a separate class independent of the score director, we can test that it behaves properly in all cases. (Environment modes are largely untested, and with the code becoming really complicated, we need to change that.)